[WIP] Task/#1372/blackoil aquifer model

kel85uk opened the pull request:
Initial commit for the Carter-Tracy analytical aquifer model. Travis CI fails for this because opm-parser has not merged the aqutab table pull requests. Otherwise, it works. We already have a skeleton structure in place.

kel85uk commented:
This PR addresses issue #1372

andlaus commented:
thanks for your contribution, it's much appreciated. This is probably be a bit nit-picky, but your PR is a bit unclean and makes reviewing it harder than necessary. could you please rebase it on top of the current master and squash the commits together?

kel85uk commented:
@andlaus I have rebased and squashed the commits. But it's still a work in progress and I need an aquifer object from parser to get it working correctly.

andlaus commented:
thanks. I can have a look at the code if you want, but verifying its correctness needs to be done by somebody who knows what he's talking about ;)

andlaus commented on opm/autodiff/AQUCT_params.hpp:
```diff
@@ -0,0 +1,31 @@
+#ifndef OPM_AQUCT_HEADER_INCLUDED
```
the copyright header is missing here. also, please to avoid underscores in filenames, i.e. rename that file to `AquctParams.cpp`.

kel85uk replied:
@andlaus Thanks for the comment, but this file is no longer in use as it was a temporary fix while I wait for the official opm-parser aquifer object. It is now done and removed.

andlaus replied:
the file is still part of the PR, right? If you don't need it could you please remove it?

kel85uk replied:
It's removed already. Thanks for the reminder. :-)

andlaus commented on opm/autodiff/BlackoilAquiferModel.hpp:
```diff
@@ -0,0 +1,181 @@
+/*
```
does this really need to be a separate class? it would probably better to integrate this into BlackOilModelEbos, so that the standard simulator supports this out of the box. Then again, maybe not...

andlaus replied:
Also, it seems like this is based on the `flow_legacy` code. are there any reasons for that?

kel85uk replied:
@andlaus I am taking the structure from the BlackoilWellModel. I am not aware which is flow_legacy style or whatnot. If we can have a discussion on this, then I would gladly change to the updated functions etc...

andlaus replied:
ah, right. I somehow messed this up. You will probably still need a separate `flow` binary if you add a new model, so it would be preferable to add aquifier support to the standard model, i.e. `BlackoilModelEbos.hpp`

kel85uk replied:
@andlaus  The aquifer model is indeed called within BlackoilModelEbos.hpp. This also means that there is an extra argument for the BlackoilModelEbos constructor and changes to the SimulatorFullyImplicitBlackoilEbos too. I anticipate that we might need such organization if we were to extend to multiple types of aquifers in the future.

andlaus replied:
I don't want to call the shots here, but I generally prefer "one size fits all" solutions because they avoid tend to mitigate combinatorial explosions of possibilities while usually only making the base code slightly more complicated.

kel85uk replied:
@andlaus Not a problem at all regarding changes, I would definitely like to hear your thoughts on how it can be organized a bit better? I am sure I am missing the point there. Is the problem with the additional arguments for the BlackoilModelEbos constructor?

andlaus replied:
Probably the issue is on my side: in my naive understanding an aquifer is just a water-only region, but based on your comments I suspect that ECL does something more crazy. is there a "two sentence executive summary" of what an ECL aquifer is?

For the issue of adding new classes, I don't mind the additional argument as I suppose the code which you propose is fine. My point is that having  a separate class for each combination of features and having N features leads to having to write 2^N classes if these features ought to be combined. even worse, these classes have to stay in sync which is IMO a direct route to maintenance hell.

Be aware that I haven't looked at the PR closely enough to determine if your "aquifer" model classes can be "bolted" on top of the regular model, in which case I would like to ask you to kindly forget my ramblings. (if so, the name of that class is a bit confusing, but again, that might be just me.) the combinatorial explosion issue is not purely theoretical rumination, though: There are already twophase, solvent and polymer (plus -- soon -- energy) extensions to flow, each of which probably ought to be combinable with aquifers.

GitPaean replied:
If there are more similar models joining, maybe we should put them in general containers based on their characteristics.

kel85uk commented:
A new commit is added which implements the equations for the carter tracy aquifer. I still need to complete the initialization routine correctly in case a user chooses the default initial aquifer pressure.

kel85uk commented:
Another WIP is the use of Eigen to make a least squares fit for the influence tables. I still need to remove the use of Eigen as I understood it will be removed in a later stage from the Opm project.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,505 @@
+/*
```
I think it is good to separate the implementation to a  `_impl.hpp`  to make this class interface easier to read.

andlaus replied:
> I think it is good to separate the implementation to a  `_impl.hpp`  to make this class interface easier to read.

I disagree because `.hpp`/`_impl.hpp`/`.cpp` splits have frequently made my life *very* miserable for stuff that I needed to modify. That said, I will not complain this (anti-) pattern for code that I'm not  supposed to touch often, but could we please agree that whether to have all function definitions in separate files or not is in the discretion of the main author of the code?

GitPaean replied:
> I disagree because .hpp/_impl.hpp/.cpp splits have frequently made my life very miserable for stuff that I needed to modify.

If you separate the definition and implementation in different files in complete way (I mean not mixed, part of implementation in `.hpp` file, part of the implementation in `_impl.hpp` file), I do not understand why you were very miserable.  Could you please explain this?   There will be some extra manual work, but comparing to the cost of learning the design/interface of a new class with 200 lines and a new class with 2000 lines, I think the extra work is small.  I know you suggested some software tools or Doxygen, but not everyone is using them and they also come with extra efforts. It is good let the code speak for themselves, like when we read a PR, what tools should we use?

kel85uk replied:
I agree with @andlaus . I had a very bad time tracking the #defines when I was starting out with the code.

akva2 replied:
keep .cpp out of the discussion. splitting into sane compile units is a *tool*, not a preference.

andlaus replied:
I think @kel85uk summarizes my troubles nicely and I do not consider myself to be a noob in this code base.

> keep .cpp out of the discussion. splitting into sane compile units is a tool, not a preference.

`.cpp` files are related to `_impl.hpp` files because if you change a non-template class to a template one you need to convert the former into the latter and this happened quite frequently in the past. another issue with this pattern are non-template classes that have some templated methods. (the latter cannot be moved to cpp, agree?)

the reason I really dislike the "separate files for the implementation" pattern is that -- besides making things  much more complicated than necessary -- it adds a lot of inertia (try to change the template parameters of such classes) and potential for error (e.g. the ordering of the template arguments can be reversed in the `_impl.hpp` file, this happened e.g. in the EQUIL init code before it was moved to eWoms) for no obvious benefit.

akva2 replied:
i said nothing about _impl files.

what i dislike is rebuilding everything and its dog just because i change a small detail. that's far far more annoying than having to relate to two editors, and is a true anti-pattern, a pattern you are advocating by including cpp files in that statement. they are, as a i said, a tool. you use it when you can, if not you do the other thing.

andlaus replied:
> what i dislike is rebuilding everything and its dog just because i change a small detail. that's far far more annoying than having to relate to two editors, and is a true anti-pattern, a pattern you are advocating by including cpp files in that statement. they are, as a i said, a tool. you use it when you can, if not you do the other thing.

This probably boils down to cultural differences: Only a single compile unit is involved for a target "everything and its dog" is still just a single compile unit. Probably we currently suffer from an "in between" approach.

It seems though that we both agree to leave the decision of how to deal with organizing code to the author of that code?

akva2 replied:
i don't agree it's a cultural thing. it's a technical thing. compile units is the chunks you can parallelize on. i like to use all my cores for building, and unless there are reasons (ie templates), one should strive for many small compile units because then we get to use our fancy multi-core machines most efficiently. everybody will have some deal with some in-between as what is done is dictated by the language and its limitations (iability to inline and such falls in this category), not personal choice.

andlaus replied:
> i don't agree it's a cultural thing. it's a technical thing.

alright, technical. you asked for it ;)

> compile units is the chunks you can parallelize on. 

I agree that this is true for C, but in C++ this dogma break down for template code (i.e., "almost always" in the case of OPM) and even if you don't use templates, the compiler being able to inline everything can have a quite large performance impact. @atgeirr learned this lesson recently, IIRC; of course most of this can amended using LTO as a band-aid, but LTO is essentially the same as using a single compile unit from the outset and prone to cause troubles in my experience.

> i like to use all my cores for building,

the same amount of parallelization can also be achieved when compiling multiple programs at once. For builds that include the tests, there are almost always enough targets.

> one should strive for many small compile units because then we get to use our fancy multi-core machines most efficiently.

I see your point, but in practice it has only limited validity: recompiling "plain ebos" is much faster when changing something in e.g. opm-material than recompilation of flow. that is due to the fact that libopmsimulators goobbles up most of the time, even when disregarding the `flow_ebos_*.cpp` CUs. If I need to recompile everything because I want to switch compilers or flags, the difference is only exacerbated.

> everybody will have some deal with some in-between as what is done is dictated by the language and its limitations (iability to inline and such falls in this category), not personal choice.

Eigen, some dune core modules and opm-material as well as eWoms do not require a library, so I think that in fact this *is* a choice that mainly depends on personal taste. Also, note that most "newer" languages like Java, Go, Rust or python do not allow a source/header split from the outset, and I think that there is a good reason for this.

GitPaean replied:
> I agree with @andlaus . I had a very bad time tracking the #defines when I was starting out with the code. The _impls wasn't very clear with the filename either for a noob such as myself...

Can you please detail the problem related to the `#defines`? Sorry I am not following this.

GitPaean replied:
> Also, note that most "newer" languages like Java, Go, Rust or python do not allow a source/header split from the outset, and I think that there is a good reason for this.  

This is not a good argument, since you probably can also argue why Java does not have `#include`, why OPM is not written in Java. 

Anyway, I am more relaxed about the `_impl.hpp` issue now, as long as the class definition layout is well made (I mean the public/private/protected member variables and member functions are well categorized in the correct location, not show up randomly). Although I still prefer a clean and shorter header file (tens of lines for a small class, a couple of hundred lines for a relatively complicated class), which makes the people learn how to use / change the new class much easier, and also it is easier to keep the class clean.  With all the features of the modern editors, I have never felt that the extra manual work to edit two files is a burden compared to other efforts involved in the development. But we can agree to disagree here. 

I agree `.cpp` file should not have been in the same discussion and that is why I did not mention it at all.

andlaus replied:
> This is not a good argument, since you probably can also argue why Java does not have #include, why OPM is not written in Java.

OPM targets the highest performance possible and this requires a language that is basically portable assembler. the fact that C++ can be used to write higher higher performance code than most other languages that were around at the start of the project does not mean all its historical features are good ideas from the conceptual point of view (the preprocessor is one of the bad ones, for example, or you can still use gotos). I guess if the project was started from scratch now, we would take a hard look at Rust an Go...

> Anyway, I am more relaxed about the _impl.hpp issue now, as long as the class definition layout is well made. [...] Although I still prefer a clean and shorter header file [...] which makes the people learn how to use / change the new class much easier.

I'd recommend to use an editor that supports code folding and to hide the method bodies if you want to get an overview...

> With all the features of the modern editors, I have never felt that the extra manual work to edit two files is a burden compared to other efforts involved in the development.

I use emacs, and switching back and forth between files is a *major* productivity killer for me. in particular, that's the case if headers are located in completely different directories than the sources or if the implementation of multiple classes is combined into a single source file. vi users -- even though they're normally the enemy ;) -- will probably agree here. if you argue that one should use a more "modern" editor, why don't you use one that supports code folding?

GitPaean replied:
> the fact that C++ can be used to write higher higher performance code than most other languages that were around at the start of the project does not mean all its historical features are good ideas from the conceptual point of view (the preprocessor is one of the bad ones, for example, or you can still use gotos).

No language is perfect, that is why the discussion should be within the scope of the specific language. 

> I use emacs, and switching back and forth between files is a major productivity killer for me. in particular, that's the case if headers are located in completely different directories than the sources or if the implementation of multiple classes is combined into a single source file.

There is no discussion about the headers in a different folder here.  I am not aware of that emacs does not support screen splitting. It is a great feature with the HD display or display with higher resolution. 

> if you argue that one should use a more "modern" editor, 

I should not have had `modern` there. Vim is a very good to go. 

Since I agree to disagree, my argument finishes here.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,505 @@
+            }
+
+            inline int
+            flowPhaseToEbosPhaseIdx( const int phaseIdx ) const
```
I think this function was already removed from the BlackoilModelEbos. I am not good about these index conversion functions, @totto82 or @andlaus can give more insight.

totto82 replied:
Yes. The code should use indices from opm-material and ewoms directly. The only part of the code that still use legacy indices is related to the well code.

kel85uk replied:
Ah nice! I was wondering about so much of those boilerplate codes... All set then. Will remove it in the cleanup commits. 👍

GitPaean replied:
There are quiet some things are defined while not used. For example, for this model, you do not introduce new primary variables or extra equations.  Some cleaning up of unused stuff will be very helpful.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,505 @@
+            inline void get_current_Pressure_cell(std::vector<Scalar>& pressure_water, const int idx, const IntensiveQuantities& intQuants)
+            {
+                const auto& fs = intQuants.fluidState();
+                pressure_water[idx] = fs.pressure(FluidSystem::waterPhaseIdx).value();
```
If you think you are solving the Aquifer related in a implicit way, you should probably keep the derivatives here, which makes the pressure_water be the variable of type of `Eval`.  Please keep in mind, with the current local AD, the derivatives are against the primary variables related to the this specific cell only, they should not interact with derivatives from other cells.                                                                     

Solving with explicit way might work as long as it gets converged.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,505 @@
+            inline void get_current_density_cell(std::vector<Scalar>& rho_water, const int idx, const IntensiveQuantities& intQuants)
+            {
+                const auto& fs = intQuants.fluidState();
+                rho_water[idx] = fs.density(FluidSystem::waterPhaseIdx).value();
```
You can think if you want to keep the derivatives from the water density, it depends on the formulation you are actually using.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,505 @@
+                Scalar a, b;
+                calculate_a_b_constants(a,b,idx,timer);
+                // This function implements Eq 5.7 of the EclipseTechnicalDescription
+                Qai_[idx] = area_fraction(idx)*( a - b * ( pressure_current_[idx] - pressure_previous_[idx] ) );
```
Based on this formulation, probably the derivatives of pressure_current_ should be kept.

nairr commented:
Thanks for the tips! I am now getting a derivative for Qai_ by setting pressure_current_ and water_density to type Eval.

However, Qai_ is a function of water phase pressure of the grid block the aquifer is connected to and the derivative im getting currently is w.r.t the water phase pressure (I think) . 
- Is it correct to update the jacobian with the derivative of Qai_  w.r.t the water phase pressure?
- Or should I be updating the jacobian with the derivative of Qai_ w.r.t the grid block pressure (the primary variable?)

GitPaean commented:
> However, Qai_ is a function of water phase pressure of the grid block the aquifer is connected to and the derivative im getting currently is w.r.t the water phase pressure (I think) .

When you get the pressure_current_, it carries the value of the water pressure and derivatives of water pressure against the primary variables (oil pressure?, water saturation and another primary variable depending on the state of the cell). Same with the water_density. When you calculate Qai_ with pressure_current and water_density, you get the value of Qai_ and the derivative of Qai_ against the primary variables (oil pressure? water saturation and ... ). You use the derivative of Qai_ against the primary variables to update the Jacobian, which carries the derivatives of all the mass balance equations against the primary variables. 

> Is it correct to update the jacobian with the derivative of Qai_ w.r.t the water phase pressure?
> Or should I be updating the jacobian with the derivative of Qai_ w.r.t the grid block pressure (the primary variable?)

The primary variables have stayed the same during the calculation. You have never calculated the derivatives against water pressure explicitly. It is handled implicitly during the calculation, basically the chain rules.

GitPaean commented:
Please update and let us know when the PR is ready for review.

kel85uk commented:
@GitPaean Yes, the initialization part is done, and the Eigen dependency has been removed. But still need to test it out, should be done by this coming Wednesday.

kel85uk commented:
@GitPaean I think you can start the review process. Some preliminary testing results suggest it is ok to proceed.

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1073,6 +1094,26 @@ namespace Opm {
+        const BlackoilAquiferModel<TypeTag>&
+        aquiferModel() const { return aquifer_model_; }
+
+        int ebosPhaseToFlowCanonicalPhaseIdx( const int phaseIdx ) const
```
Is this function used in any way? If not, please remove it.

kel85uk replied:
@GitPaean  I have removed this. Thanks

GitPaean commented on opm/autodiff/BlackoilAquiferModel.hpp:
```diff
@@ -0,0 +1,163 @@
+*/
+
+
+#ifndef OPM_BLACKOILAQUIFERMODEL_HEADER_INCLUDED
```
Same applies to this class. And if a function is empty or not used, please remove them.

kel85uk replied:
@GitPaean Likewise too.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,443 @@
+            static const int Oil = BlackoilPhases::Liquid;
+            static const int Gas = BlackoilPhases::Vapour;
+
+            typedef typename GET_PROP_TYPE(TypeTag, Grid) Grid;
```
This class requires some substantial cleaning up. Basically, I guess the Oil, Gas, and polymer and solvent related are not needed in the current pull request, please remove them.

GitPaean commented on opm/autodiff/BlackoilAquiferModel.hpp:
```diff
@@ -0,0 +1,131 @@
+            typedef typename GET_PROP_TYPE(TypeTag, ElementContext)      ElementContext;
+            typedef typename GET_PROP_TYPE(TypeTag, Indices)             BlackoilIndices;
+            typedef typename GET_PROP_TYPE(TypeTag, Simulator)           Simulator;
+            typedef typename GET_PROP_TYPE(TypeTag, Scalar)              Scalar;
```
Only `Scalar`, `Simulator`, `ElementContext` are required here, please remove the unused types here.

GitPaean commented on opm/autodiff/BlackoilAquiferModel.hpp:
```diff
@@ -0,0 +1,131 @@
+            }
+
+            // This initialization function is used to connect the parser objects with the ones needed by AquiferCarterTracy
+            void init(const Simulator& ebosSimulator, std::vector<Aquifer_object>& aquifers);
```
Both `ebosSimuator` and `aquifers` are member variable, they do not need to function arguments.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+        public:
+            typedef BlackoilModelParameters ModelParameters;
+
+            typedef typename GET_PROP_TYPE(TypeTag, Grid) Grid;
```
Please also remove the unused types here.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+            :  ebos_simulator_ (ebosSimulator),
+              aquiferID_ (params.aquiferID),
+              inftableID_ (params.inftableID),
+              pvttableID_ (params.pvttableID),
```
Please add `AQUCT_data` as a member (or reference if appropriate) to reduce the complexity of the class `AquiferCarterTracy`.  And `params` is a bad name of this argument, how about `aquct_data` or something similar.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+            }
+
+
+            inline const std::vector<int> cell_id() const
```
These two functions are not called.  Please remove them.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+                init_quantities(connection);
+            }
+
+            inline void assembleAquiferEq(Simulator& ebosSimulator, const SimulatorTimerInterface& timer)
```
Since you have `ebos_simulator_` as a member variable, `ebosSimulator` is probably not needed here.

GitPaean commented on opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp:
```diff
@@ -187,6 +191,8 @@ class SimulatorFullyImplicitBlackoilEbos
             ebosSimulator_.model().addAuxiliaryModule(auxMod);
         }
 
+        AquiferModel aquifer_model(ebosSimulator_, model_param_, terminal_output_);
```
A question here. Does `Aquifer` use any results from the `Aquifer` in the previous report step? I mean, between report steps, does any information/results from previous time step transfer to the current time step? If not, it is possible the aquifer model can be created within the `BlackoilModelEbos` directly without being an argument pass into `BlackoilModelEbos`. Then the changes in this file should be removed.

GitPaean replied:
Now I see `W_flux_` is one thing transferred between time steps.

GitPaean commented on opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp:
```diff
@@ -29,6 +29,9 @@
 #include <opm/autodiff/WellStateFullyImplicitBlackoil.hpp>
 #include <opm/autodiff/BlackoilWellModel.hpp>
+#include <opm/autodiff/BlackoilAquiferModel.hpp>
+#include <opm/autodiff/RateConverter.hpp>
```
I am not sure why adding these two header files here. 
```
+#include <opm/autodiff/RateConverter.hpp>
+#include <opm/autodiff/SimFIBODetails.hpp>
```

GitPaean commented on opm/autodiff/BlackoilAquiferModel.hpp:
```diff
@@ -0,0 +1,131 @@
+
+            void updateConnectionIntensiveQuantities() const;
+
+            int numAquifers() const;
```
This function is not used any where, need to be removed.

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -29,6 +29,9 @@
 #include <opm/autodiff/BlackoilModelParameters.hpp>
 #include <opm/autodiff/BlackoilWellModel.hpp>
+#include <opm/autodiff/BlackoilAquiferModel.hpp>
+#include <opm/autodiff/GridHelpers.hpp>
```
Please make sure 

```
+#include <opm/autodiff/GridHelpers.hpp>
+#include <opm/autodiff/GeoProps.hpp>
```

are needed.

GitPaean commented on opm/autodiff/BlackoilAquiferModel_impl.hpp:
```diff
@@ -0,0 +1,138 @@
+    }
+
+
+    template<typename TypeTag>
```
What is the function `updateConnectionIntensiveQuantities` doing? Could you please provide some explanation here? I am not familiar with `ElementContext` related.  Maybe @andlaus can provide some light here.

kel85uk replied:
@GitPaean Yes, it's supposed to get the updated solution at the same iteration, this is required so that the aquifer model can get the most recent solution field connected to the model.

As a very brief background, the model implements the inflow, Q as a function of the pressure field, P, which is implicit.

$$Q_{ai} = f(p_i(t+\delta t) - p_i(t))$$

GitPaean commented on opm/autodiff/BlackoilAquiferModel_impl.hpp:
```diff
@@ -0,0 +1,138 @@
+
+        // number_of_cells_ = gridView.size(/*codim=*/0);
+        // global_nc_ = gridView.comm().sum(number_of_cells_);
+        gravity_ = ebosSimulator_.problem().gravity()[2];
```
Since ebosSimulator_ is a member, gravity_ does not need to be a member variable here.

GitPaean commented on opm/autodiff/BlackoilAquiferModel.hpp:
```diff
@@ -0,0 +1,131 @@
+            // called at the end of a time step
+            void timeStepSucceeded(const SimulatorTimerInterface& timer);
+
+            inline const Simulator& simulator() const
```
This function is not used.

GitPaean commented on opm/autodiff/BlackoilAquiferModel.hpp:
```diff
@@ -0,0 +1,131 @@
+#include <opm/common/utility/platform_dependent/disable_warnings.h>
+#include <opm/common/utility/platform_dependent/reenable_warnings.h>
+
+#include <cassert>
```
Please remove the header files not used.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+            inline Scalar calculate_reservoir_equilibrium(std::vector<Eval>& rho_water_reservoir)
+            {
+                // Since the global_indices are the reservoir index, we just need to extract the fluidstate at those indices
+                std::vector<Scalar> water_pressure_reservoir, pw_aquifer;
```
`water_pressure_reservoir` does not need to be a vector here. It can be one value for each `for` step.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+                
+                if (p0_defaulted_)
+                {
+                   pa0_ = calculate_reservoir_equilibrium(rhow_);
```
rhoW_ is a member variable, it does not need to be an argument of the member function.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+                        // the logically-Cartesian direction of the face
+                        const int faceTag = Opm::UgGridHelpers::faceTag(ugrid, cellFaceIter);
+
+                        
```
Please use switch and also provide a `default` option (probably throw a run-time error. ) In the current form, `faceDirection` has the risk of not being initialized.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+                auto cell2Faces = Opm::UgGridHelpers::cell2Faces(ugrid);
+                auto faceCells  = Opm::AutoDiffGrid::faceCells(ugrid);
+
+                for (auto influxCoeff: connection.influx_coeff){
```
Please remove the  debugging output.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,418 @@
+                Qai_.resize(cell_idx_.size(), 0.0);
+            }
+
+            inline void get_current_Pressure_cell(std::vector<Eval>& pressure_water, const int idx, const IntensiveQuantities& intQuants)
```
please change the name of the function. You use the function get_current_Pressure_cell to set both the `pressure_current_` and `pressure_previous_`, which is really confusing. You can change the function name to be
something like `set_pressure_cell` or something. The function is more like doing `setting` other than `getting`.

kel85uk commented on opm/autodiff/BlackoilAquiferModel_impl.hpp:
```diff
@@ -0,0 +1,138 @@
+        std::vector<AquiferCT::AQUCT_data> aquifersData = aquiferct.getAquifers();
+        std::vector<Aquancon::AquanconOutput> aquifer_connection = aquifer_connect.getAquOutput();
+
+        assert( aquifersData.size() == aquifer_connect.size() );
```
@GitPaean Hi, it's strange that this line compiles in my system. But it is definitely not correct, as I intended it to be aquifer_connection instead of aquifer_connect. Thanks!

andlaus replied:
> it's strange that this line compiles in my system.

you probably compiled it in release mode or with NDEBUG defined. asserts will be completely ignored in that case.

atgeirr replied:
> Hi, it's strange that this line compiles in my system.

If you only compile a Release build, you will not see any errors. This is because assert is a macro, that expands to nothing if NDEBUG is set, which is true for Release mode.

It is recommended to make a debug build as well, to trigger asserts when testing.

kel85uk replied:
Ah nice. Thanks for the information.

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -362,9 +368,22 @@ namespace Opm {
+                // Modify the Jacobian and residuals according to the aquifer models
+                aquiferModel().assemble(timer, iterationIdx);
+            }
+            catch( const Dune::FMatrixError& e )
```
You can do that, while I am not sure what is the correct thing to do here. Should we re-throw `Opm::NumericalIssue` or `std::runtime_error` here? @atgeirr , any comments about this issue?

atgeirr replied:
I think it is ok to catch(...) here and throw again, it helps to point out where the error happened.

kel85uk commented:
@GitPaean Most if not all cleanups are done. Thanks.

GitPaean commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,362 @@
+            }
+
+            // This function implements Eq 5.7 of the EclipseTechnicalDescription       
+            inline void calculate_inflow_rate(int idx, const SimulatorTimerInterface& timer)
```
If you would like to, you can use camelCase for the functions. It is more consistent with the rest part of OPM.

GitPaean commented:
I believe this PR is dependent on https://github.com/OPM/opm-common/pull/377 , it should wait until https://github.com/OPM/opm-common/pull/377 gets merged first.

GitPaean commented:
> @GitPaean Hi, I just found out that the PR from @nairr for opm-common will be merged today?

I am not sure, but that PR can be merged independent of this PR. 

And also, please address the remaining few comments and try to create a test case with aquifer (maybe the one you guys created based on one case I shared with you, since it is publicly available) in opm-data. I am not sure how to create a new test case with `opm-tests` yet, any suggestion for the correct procedure, @akva2 ?

GitPaean commented:
@kel85uk , if possible, you can also rebase the branch to the current master branch.

GitPaean commented:
jenkins build this please

GitPaean commented:
Please address the compilation failure.

kel85uk commented on opm/autodiff/AquiferCarterTracy.hpp:
```diff
@@ -0,0 +1,362 @@
+            std::vector<Scalar> alphai_;
+
+            // Variables constants
+            const AquiferCT::AQUCT_data aquct_data_;
```
@GitPaean Somehow, referencing this makes it to not get the correct values. I will need to check this more carefully.

kel85uk commented:
jenkins build this please

kel85uk commented:
@GitPaean Hi, Have done the cleaning up based on your comments. Except for the reference of aquct_data_.

GitPaean commented:
jenkins build this please

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -409,13 +426,13 @@ namespace Opm {
                     continue;
 
-		unsigned globalElemIdx = elemMapper.index(elem);
+        unsigned globalElemIdx = elemMapper.index(elem);
```
Anyone has some idea about what is going on about these indentations here?  TAB or Space related issue?

kel85uk replied:
@GitPaean It seems like it. My apologies. Seems like one alignment out (4 space) tab.

GitPaean replied:
Please recover the unnecessary changes in BlackoilModelEbos.hpp, like these indentation and also the above `const double dt = timer.currentStepLength();`. Basically, any change should have a good reason hopefully.

kel85uk commented:
@GitPaean Done. Can we start the merging process already?

kel85uk commented:
@nairr As requested from @GitPaean , we do need to add the integration test.

GitPaean commented:
jenkins build this please

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -366,8 +370,21 @@ namespace Opm {
             ebosSimulator_.problem().endIteration();
 
+            // -------- Current time step length ----------
+            const double dt = timer.currentStepLength();
```
Please move these two lines to the original place, before Well equations?, since it is closer to where `dt` is used.

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -579,7 +596,6 @@ namespace Opm {
           virtual void apply( const X& x, Y& y ) const
           {
             A_.mv( x, y );
-
```
remove this deletion please.

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -593,7 +609,6 @@ namespace Opm {
           virtual void applyscaleadd (field_type alpha, const X& x, Y& y) const
           {
             A_.usmv(alpha,x,y);
-
```
remove this deletion please.

GitPaean commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1100,6 +1118,12 @@ namespace Opm {
+        BlackoilAquiferModel<TypeTag>&
+        aquiferModel() { return aquifer_model_; }
+
+        const BlackoilAquiferModel<TypeTag>&
```
This const function might not be needed/used.

GitPaean commented:
A few very minor comments, then my review is done.  If possible, please incorporate https://github.com/OPM/opm-simulators/pull/1487 to this PR, so we can do a final jenkins test, with `update_data`. Thanks for the contribution.

kel85uk commented:
@GitPaean Hi, I have already made the changes requested. Also, merged the branch of PR #1487, which can be closed then too.

nairr commented:
@kai could you please invoke jenkins update_data again ?

nairr commented:
@GitPaean could you please invoke jenkins update_data again?

kel85uk commented:
Question to the community: How do we even call jenkins to test our own commits?

blattms commented:
See the wiki at https://wiki.opm-project.org/index.php?title=Jenkins_triggers

But you have to have permissions to be heard by jenkins.

alfbr commented:
But to test your own commits you can simply run the tests locally on your machine, downloading the opm-tests repo.

GitPaean commented on tests/update_reference_data.sh:
```diff
@@ -102,7 +102,7 @@ for test_name in ${tests}; do
     copyToReferenceDir \
       $configuration/build-opm-simulators/tests/results/flow+ctaquifer_2d_oilwater/ \
-      $OPM_TESTS_ROOT/aquifer-oilwater/opm-simulation-reference/flow \
+      $OPM_TESTS_ROOT/aquifer-oilwater/eclipse-simulation \
```
Sorry, but I am not sure this makes sense. @akva2 , any comment?

GitPaean replied:
Integration tests testing against the reference result under `opm-simulation-reference`, we need to make sure the `opm-simulation-reference` always agrees well with the `eclipse-simulation` result by ourselves, while we do not compare the results against the `eclipse-simulation` during integration tests. This is my understanding.

GitPaean commented:
> But to test your own commits you can simply run the tests locally on your machine, downloading the opm-tests repo.

Yes. We are supposed to do local tests and then push. Although I did not do that for a few times, like once this morning, I thought was a trivial code without testing locally and jenkins failed in compilation.

GitPaean commented:
jenkins build this update_data please

GitPaean commented:
jenkins build this opm-tests=20 please

GitPaean commented:
One of the NOTEs about this PR is, it does not support parallel running of the Aquifer model. The information of the Aquifer in the parser is not partitioned, any ideas about where we should do the partitioning? I guess is when we create the BlackoilAquiferModel.  More specifically, is that in the `init()` function of `BlackoilAquiferModel`.  With `ebosSimulator_`, would we have enough information to do the partitioning of the partitioning? I am okay to put this in another PR. Any comments, @andlaus @blattms @atgeirr @alfbr ?

GitPaean commented:
I am doing the final tests with some of the big models to see the overhead of the model on the running without aquifer models.

GitPaean commented:
It has several percent of extra overload for running of Norne. @kel85uk , I proposed some change to reduce this part to your branch, please take look at it or suggest some other better way. 

The change introduced in https://github.com/OPM/opm-simulators/pull/1371/commits/dc56b967c38461dbdce9a90087ddc51ba5383def  is not desirable, please remove the changes.

andlaus commented:
about the parallelization: I am not familiar enough with the present PR to give any good advice how to do this, but I agree @GitPaean's assessment that only having aquifer support for sequential runs is sufficient for now. If the results of parallel runs with aquifers are wrong -- instead of just slow -- it would IMO be good to add an error message and abort the simulator for parallel runs with aquifer-using decks as a safeguard. (I don't know if this is already the case.)

GitPaean commented:
> If the results of parallel runs with aquifers are wrong -- instead of just slow -- it would IMO be good to add an error message and abort the simulator for parallel runs with aquifer-using decks as a safeguard. 

It just gives `segmentation fault` for now. I agree we should  give some more meaningful message for this. 
In the `SimulatorFullyImplicitBlackoilEbos`, there is  a bool `is_parallel_run_`, we can incorporate this information to the constructor of `BlackoilAquiferModel`. When there is keyword `AQUCT` (not sure this is the best keywords for the indication of including CT aquifer), and it is parallel running, we should throw a run-time error and stop the running.

kel85uk commented:
@GitPaean I have removed commit dc56b967c38461dbdce9a90087ddc51ba5383def.

GitPaean commented:
> @GitPaean I have removed commit dc56b96.

Hi, it looks like you also removed the things introduced in https://github.com/kel85uk/opm-simulators/pull/13 . I am not sure how you removed dc56b96, but be careful not remove what you have developed by accident.

GitPaean commented:
jenkins build this opm-tests=20 please

GitPaean commented:
jenkins build this update_data please

GitPaean commented:
The major remaining issue is the support of the parallel running for the aquifer model.   The result of the aquifer case in opm-test agrees well with the reference result.  The impact on the performance of running models without aquifer model is minimal.  Jenkins is happy with the updated reference of the aquifer model in OPM/opm-tests#22 . 

Thanks for the contribution and patience. Merging.
